src/decima/data/dataset.py

```diff
@@ -0,0 +1,365 @@
+            }
+        )
+        df["rel_pos"] = df.apply(VariantDataset._relative_pos, axis=1)
+        df["ref_tx"] = df.apply(
+            lambda row: row.ref if row.strand == "+" else reverse_complement(row.ref),
+            axis=1,
+        )
+        df["alt_tx"] = df.apply(
+            lambda row: row.alt if row.strand == "+" else reverse_complement(row.alt),
+            axis=1,
+        )
+        df["tss_dist"] = df.rel_pos - df.gene_mask_start
+
+        df = df[(df.rel_pos > min_from_end) & (df.rel_pos < DECIMA_CONTEXT_SIZE - min_from_end)]
+
+        if distance_type == "tss":
+            df = df[(df.tss_dist.abs() >= min_distance) & (df.tss_dist.abs() < max_distance)]
+        elif distance_type == "gene":
+            raise NotImplementedError("max_distance_type == 'gene' is not implemented yet.")
+        else:
+            raise ValueError(f"Invalid distance_type: {distance_type}. Must be 'tss'.")
+
+        if df.shape[0] == 0:
+            raise ValueError(
+                "NoOverlapError: There is no overlap between provided variants and genes."
+                " Check `max_distance` and `min_distance` arguments."
+            )
+
+        return df[[*VariantDataset.DEFAULT_COLUMNS, *include_cols]]
+
+    @staticmethod
+    def _relative_pos(row):
+        # + 1 because gene start is 0 based, variant pos is 1 based and gene end is 1 based
+        # TO CONSIDER: relative position is not valid for indels because of the shifting of the sequence.
+        # so relative position of reference allele and alterantive allele is different
+        rel_pos = row.pos - (row.start + 1) if row.strand == "+" else row.end - row.pos
+        return rel_pos
+
+    def _overlap_genes(
+        self,
+        variants,
+        include_cols=None,
+        gene_col=None,
+        min_from_end=0,
+        distance_type="tss",
+        min_distance=0,
+        max_distance=float("inf"),
+    ):
+        return self.overlap_genes(
+            variants,
+            self.result.gene_metadata.reset_index(names=["gene"]),
+            include_cols=include_cols,
+            gene_col=gene_col,
+            min_from_end=min_from_end,
+            distance_type=distance_type,
+            min_distance=min_distance,
+            max_distance=max_distance,
+        )
+
+    def __len__(self):
+        return self.n_seqs * self.n_augmented * self.n_alleles
+
+    def validate_allele_seq(self, gene, variant):
+        seq = self.result.gene_sequence(gene)
+        vstart = variant.rel_pos
+        vend = vstart + len(variant.ref)
+        return (seq[vstart:vend] == variant.ref_tx) or (seq[vstart:vend] == variant.alt_tx)
+
+    def __getitem__(self, idx):
+        seq_idx, augment_idx, allele_idx = _split_overall_idx(idx, (self.n_seqs, self.n_augmented, self.n_alleles))
+
+        variant = self.variants.iloc[seq_idx]
+
+        warnings = list()
+        if not self.validate_allele_seq(variant.gene, variant):
```

Comment:
Why not check the allele mismatch only if allele_idx = True? In that case, it will only run when the alternate allele is processed, instead of running twice per variant, and you won't have to divide the number of warnings by 2 later.